File: NovilhaNutri/Core/Calculo/RelatorioTexto.swift
import Foundation

/// Gera a versão em texto do planejamento, para compartilhar ou imprimir.
enum RelatorioTexto {

    static func gerar(_ relatorio: RelatorioPlanejamento) -> String {
        let lote = relatorio.lote
        var linhas: [String] = []

        linhas.append("PLANEJAMENTO NUTRICIONAL E DE ABATE")
        linhas.append(String(repeating: "=", count: 42))
        linhas.append("Lote: \(lote.nome)")
        linhas.append("Animais: \(relatorio.animais) cabeças")
        linhas.append("Categoria: novilhas - \(lote.fase.nome) - \(lote.grupoGenetico.nomeCurto)")
        linhas.append("Sistema: \(lote.sistema.nome)")
        linhas.append("")

        guard relatorio.viavel else {
            linhas.append("Sem projeção disponível.")
            linhas.append(contentsOf: relatorio.alertas.map { "- " + $0 })
            return linhas.joined(separator: "\n")
        }

        linhas.append("METAS")
        linhas.append("Peso atual: \(Formatadores.kg(relatorio.pesoInicial))")
        linhas.append("Peso alvo de abate: \(Formatadores.kg(relatorio.pesoAlvo))")
        linhas.append("Ganho médio diário: \(Formatadores.numero(lote.ganhoMetaDiario, casas: 3)) kg/dia")
        linhas.append("Período: \(Formatadores.duracao(dias: relatorio.diasTotais)) (\(Formatadores.numero(relatorio.diasTotais, casas: 0)) dias)")
        linhas.append("Início: \(Formatadores.data(relatorio.dataInicio))")
        linhas.append("Abate previsto: \(Formatadores.data(relatorio.dataAbate))")
        linhas.append("")

        linhas.append("PRODUÇÃO PREVISTA")
        linhas.append("Ganho por animal: \(Formatadores.kg(relatorio.ganhoPorAnimal))")
        linhas.append("Ganho do lote: \(Formatadores.kg(relatorio.ganhoTotalLote))")
        linhas.append("Rendimento de carcaça: \(Formatadores.percentual(lote.rendimentoCarcaca * 100))")
        linhas.append("Carcaça por animal: \(Formatadores.kg(relatorio.pesoCarcacaFinal)) (\(Formatadores.arroba(relatorio.arrobasFinais)))")
        linhas.append("Arrobas produzidas por animal: \(Formatadores.arroba(relatorio.arrobasProduzidasPorAnimal))")
        linhas.append("Arrobas produzidas no lote: \(Formatadores.arroba(relatorio.arrobasProduzidasLote))")
        linhas.append("Arrobas totais no abate: \(Formatadores.arroba(relatorio.arrobasTotaisLote))")
        linhas.append("Conversão alimentar: \(Formatadores.numero(relatorio.conversaoAlimentar)) kg MS por kg de ganho")
        linhas.append("")

        linhas.append("INSUMOS DO CICLO COMPLETO")
        for total in relatorio.totais {
            linhas.append("- \(total.insumo.nome)")
            linhas.append("    \(Formatadores.numero(total.kgMateriaNatural, casas: 0)) kg naturais (\(Formatadores.numero(total.toneladas, casas: 2)) t)")
            if let conversao = total.conversao {
                linhas.append("    \(conversao.descricao) - comprar \(conversao.descricaoCompra)")
            } else {
                linhas.append("    fornecido no pastejo")
            }
            if total.custo > 0 {
                linhas.append("    custo \(Formatadores.moeda(total.custo))")
            }
        }
        linhas.append("")

        if relatorio.custoTotal > 0 {
            linhas.append("CUSTOS")
            linhas.append("Custo total do ciclo: \(Formatadores.moeda(relatorio.custoTotal))")
            linhas.append("Custo por animal: \(Formatadores.moeda(relatorio.custoPorAnimal))")
            linhas.append("Custo por animal por dia: \(Formatadores.moeda(relatorio.custoPorAnimalDia))")
            linhas.append("Custo por quilo ganho: \(Formatadores.moeda(relatorio.custoPorKgGanho))")
            linhas.append("Custo por arroba produzida: \(Formatadores.moeda(relatorio.custoPorArroba))")
            linhas.append("")
        }

        if relatorio.temPrecos {
            let modo = lote.modoCompra
            linhas.append("RESULTADO PREVISTO")
            var compra = "Compra: \(Formatadores.moeda(lote.precoCompra)) \(modo.porQue)"
            if modo == .porArroba {
                compra += " (\(Formatadores.arroba(lote.arrobasCompra)))"
            }
            linhas.append(compra)
            linhas.append("Venda: \(Formatadores.moeda(lote.precoArrobaVenda)) por arroba")
            linhas.append("")
            linhas.append("Compra por animal: \(Formatadores.moeda(relatorio.compraPorAnimal))")
            linhas.append("Dieta por animal: \(Formatadores.moeda(relatorio.custoPorAnimal))")
            linhas.append("Investido por animal: \(Formatadores.moeda(relatorio.investimentoPorAnimal))")
            linhas.append("Venda por animal: \(Formatadores.moeda(relatorio.receitaPorAnimal))")
            linhas.append("Lucro por animal: \(Formatadores.moeda(relatorio.lucroPorAnimal))")
            linhas.append("")
            linhas.append("Compra do lote: \(Formatadores.moeda(relatorio.compraTotal))")
            linhas.append("Dieta do lote: \(Formatadores.moeda(relatorio.custoTotal))")
            linhas.append("Investido no lote: \(Formatadores.moeda(relatorio.investimentoTotal))")
            linhas.append("Venda do lote: \(Formatadores.moeda(relatorio.receitaTotal))")
            linhas.append("Lucro do lote: \(Formatadores.moeda(relatorio.lucroTotal))")
            linhas.append("")
            linhas.append("Margem sobre a venda: \(Formatadores.percentual(relatorio.margemSobreReceita))")
            linhas.append("Retorno sobre o investido: \(Formatadores.percentual(relatorio.retornoSobreInvestimento))")
            linhas.append("Lucro por arroba produzida: \(Formatadores.moeda(relatorio.lucroPorArrobaProduzida))")
            linhas.append("Arroba de equilíbrio: \(Formatadores.moeda(relatorio.precoArrobaEquilibrio))")
            linhas.append("Resultado só da engorda: \(Formatadores.moeda(relatorio.margemDaEngorda))")
            if lote.precoCompra <= 0 {
                linhas.append("Sem preço de compra informado: o resultado conta apenas a dieta.")
            }
            if relatorio.custoTotal <= 0 {
                linhas.append("Nenhum alimento do lote tem preço cadastrado: a dieta entra como custo zero.")
            }
            linhas.append("Não entram sanidade, transporte, pastagem, mão de obra nem impostos.")
            linhas.append("")
        }

        linhas.append("PERÍODOS")
        for periodo in relatorio.periodos {
            linhas.append("\(periodo.titulo): \(Formatadores.data(periodo.dataInicio)) a \(Formatadores.data(periodo.dataFim)) (\(Formatadores.numero(periodo.dias, casas: 0)) dias)")
            linhas.append("  Peso \(Formatadores.numero(periodo.pesoInicial, casas: 0)) a \(Formatadores.numero(periodo.pesoFinal, casas: 0)) kg")
            linhas.append("  Exigência diária por animal: MS \(Formatadores.kg(periodo.exigencia.consumoMateriaSeca)), PB \(Formatadores.gramas(periodo.exigencia.proteinaBrutaGramas)), NDT \(Formatadores.kg(periodo.exigencia.ndtKg))")
            linhas.append("  Dieta: \(Formatadores.percentual(periodo.exigencia.proteinaBrutaPercentualDieta)) PB e \(Formatadores.percentual(periodo.exigencia.ndtPercentualDieta)) NDT na matéria seca")
            for consumo in periodo.consumos {
                var texto = "  - \(consumo.insumo.nome): \(Formatadores.numero(consumo.kgMateriaNatural, casas: 0)) kg"
                if let conversao = consumo.conversao {
                    texto += " = \(conversao.descricao)"
                }
                linhas.append(texto)
            }
            if periodo.custo > 0 {
                linhas.append("  Custo do período: \(Formatadores.moeda(periodo.custo))")
            }
        }

        if !relatorio.alertas.isEmpty {
            linhas.append("")
            linhas.append("OBSERVAÇÕES")
            linhas.append(contentsOf: relatorio.alertas.map { "- " + $0 })
        }

        linhas.append("")
        linhas.append("Gerado pelo NovilhaNutri. Cálculos de referência; ajuste conforme o desempenho observado e a orientação do responsável técnico.")
        return linhas.joined(separator: "\n")
    }
}

File: web/src/nucleo/relatorioTexto.ts
/**
 * Gera a versão em texto do planejamento, para compartilhar ou imprimir.
 *
 * Porte de `Core/Calculo/RelatorioTexto.swift`, do antigo aplicativo de iPhone.
 */
import { FASES, GRUPOS, MODOS_COMPRA, SISTEMAS } from "./classificacoes.js";
import { descricao as descricaoConversao, descricaoCompra } from "./conversorSacas.js";
import {
  arroba,
  data as formatarData,
  duracao,
  gramas,
  kg as formatarKg,
  moeda,
  numero,
  percentual,
} from "./formatadores.js";
import { arrobasCompra, custoCompraPorAnimal } from "./lote.js";
import {
  arrobasFinais,
  arrobasProduzidasLote,
  arrobasProduzidasPorAnimal,
  consumoConversao,
  conversaoAlimentar,
  custoPorAnimal,
  custoPorAnimalDia,
  custoPorArroba,
  custoPorKgGanho,
  custoPeriodo,
  custoTotal,
  compraTotal,
  ganhoPorAnimal,
  ganhoTotalLote,
  investimentoPorAnimal,
  investimentoTotal,
  lucroPorAnimal,
  lucroPorArrobaProduzida,
  lucroTotal,
  margemDaEngorda,
  margemSobreReceita,
  pesoCarcacaFinal,
  precoArrobaEquilibrio,
  receitaPorAnimal,
  receitaTotal,
  resumosPorEtapa,
  retornoSobreInvestimento,
  temDuasEtapas,
  temPrecos,
  tituloPeriodo,
  viavel,
  consumoToneladas,
  type RelatorioPlanejamento,
} from "./planejadorAbate.js";

export function gerar(relatorio: RelatorioPlanejamento): string {
  const lote = relatorio.lote;
  const linhas: string[] = [];

  linhas.push("PLANEJAMENTO NUTRICIONAL E DE ABATE");
  linhas.push("=".repeat(42));
  linhas.push(`Lote: ${lote.nome}`);
  linhas.push(`Animais: ${relatorio.animais} cabeças`);
  linhas.push(
    `Categoria: novilhas - ${FASES[lote.fase].nome} - ${GRUPOS[lote.grupoGenetico].nomeCurto}`,
  );
  linhas.push(`Sistema: ${SISTEMAS[lote.sistema].nome}`);
  linhas.push("");

  if (!viavel(relatorio)) {
    linhas.push("Sem projeção disponível.");
    linhas.push(...relatorio.alertas.map((a) => `- ${a}`));
    return linhas.join("\n");
  }

  linhas.push("METAS");
  linhas.push(`Peso atual: ${formatarKg(relatorio.pesoInicial)}`);
  linhas.push(`Peso alvo de abate: ${formatarKg(relatorio.pesoAlvo)}`);
  linhas.push(`Ganho médio diário: ${numero(lote.ganhoMetaDiario, 3)} kg/dia`);
  linhas.push(
    `Período: ${duracao(relatorio.diasTotais)} (${numero(relatorio.diasTotais, 0)} dias)`,
  );
  linhas.push(`Início: ${formatarData(relatorio.dataInicio)}`);
  linhas.push(`Abate previsto: ${formatarData(relatorio.dataAbate)}`);
  linhas.push("");

  linhas.push("PRODUÇÃO PREVISTA");
  linhas.push(`Ganho por animal: ${formatarKg(ganhoPorAnimal(relatorio))}`);
  linhas.push(`Ganho do lote: ${formatarKg(ganhoTotalLote(relatorio))}`);
  linhas.push(`Rendimento de carcaça: ${percentual(lote.rendimentoCarcaca * 100)}`);
  linhas.push(
    `Carcaça por animal: ${formatarKg(pesoCarcacaFinal(relatorio))} ` +
      `(${arroba(arrobasFinais(relatorio))})`,
  );
  linhas.push(
    `Arrobas produzidas por animal: ${arroba(arrobasProduzidasPorAnimal(relatorio))}`,
  );
  linhas.push(`Arrobas produzidas no lote: ${arroba(arrobasProduzidasLote(relatorio))}`);
  linhas.push(`Arrobas totais no abate: ${arroba(arrobasFinais(relatorio) * relatorio.animais)}`);
  linhas.push(
    `Conversão alimentar: ${numero(conversaoAlimentar(relatorio))} kg MS por kg de ganho`,
  );
  linhas.push("");

  if (temDuasEtapas(relatorio)) {
    linhas.push("ETAPAS DA DIETA");
    for (const etapa of resumosPorEtapa(relatorio)) {
      linhas.push(
        `${etapa.nome}: ${formatarKg(etapa.pesoInicial)} a ${formatarKg(etapa.pesoFinal)}, ` +
          `${numero(etapa.ganhoDiario, 3)} kg/dia, ${numero(etapa.dias, 0)} dias`,
      );
      for (const total of etapa.totais) {
        const conversao = consumoConversao(total);
        let texto = `  - ${total.insumo.nome}: ${numero(total.kgMateriaNatural, 0)} kg`;
        if (conversao) texto += ` = ${descricaoCompra(conversao)}`;
        linhas.push(texto);
      }
      if (etapa.custo > 0) linhas.push(`  Custo da etapa: ${moeda(etapa.custo)}`);
    }
    linhas.push("");
  }

  linhas.push("INSUMOS DO CICLO COMPLETO");
  for (const total of relatorio.totais) {
    linhas.push(`- ${total.insumo.nome}`);
    linhas.push(
      `    ${numero(total.kgMateriaNatural, 0)} kg naturais ` +
        `(${numero(consumoToneladas(total), 2)} t)`,
    );
    const conversao = consumoConversao(total);
    if (conversao) {
      linhas.push(
        `    ${descricaoConversao(conversao)} - comprar ${descricaoCompra(conversao)}`,
      );
    } else {
      linhas.push("    fornecido no pastejo");
    }
    if (total.custo > 0) linhas.push(`    custo ${moeda(total.custo)}`);
  }
  linhas.push("");

  if (custoTotal(relatorio) > 0) {
    linhas.push("CUSTOS");
    linhas.push(`Custo total do ciclo: ${moeda(custoTotal(relatorio))}`);
    linhas.push(`Custo por animal: ${moeda(custoPorAnimal(relatorio))}`);
    linhas.push(`Custo por animal por dia: ${moeda(custoPorAnimalDia(relatorio))}`);
    linhas.push(`Custo por quilo ganho: ${moeda(custoPorKgGanho(relatorio))}`);
    linhas.push(`Custo por arroba produzida: ${moeda(custoPorArroba(relatorio))}`);
    linhas.push("");
  }

  if (temPrecos(relatorio)) {
    const modo = MODOS_COMPRA[lote.modoCompra];
    linhas.push("RESULTADO PREVISTO");
    linhas.push(
      `Compra: ${moeda(lote.precoCompra)} ${modo.porQue}` +
        (lote.modoCompra === "porArroba" ? ` (${arroba(arrobasCompra(lote))})` : ""),
    );
    linhas.push(`Venda: ${moeda(lote.precoArrobaVenda)} por arroba`);
    linhas.push("");
    linhas.push(`Compra por animal: ${moeda(custoCompraPorAnimal(lote))}`);
    linhas.push(`Dieta por animal: ${moeda(custoPorAnimal(relatorio))}`);
    linhas.push(`Investido por animal: ${moeda(investimentoPorAnimal(relatorio))}`);
    linhas.push(`Venda por animal: ${moeda(receitaPorAnimal(relatorio))}`);
    linhas.push(`Lucro por animal: ${moeda(lucroPorAnimal(relatorio))}`);
    linhas.push("");
    linhas.push(`Compra do lote: ${moeda(compraTotal(relatorio))}`);
    linhas.push(`Dieta do lote: ${moeda(custoTotal(relatorio))}`);
    linhas.push(`Investido no lote: ${moeda(investimentoTotal(relatorio))}`);
    linhas.push(`Venda do lote: ${moeda(receitaTotal(relatorio))}`);
    linhas.push(`Lucro do lote: ${moeda(lucroTotal(relatorio))}`);
    linhas.push("");
    linhas.push(`Margem sobre a venda: ${percentual(margemSobreReceita(relatorio))}`);
    linhas.push(`Retorno sobre o investido: ${percentual(retornoSobreInvestimento(relatorio))}`);
    linhas.push(`Lucro por arroba produzida: ${moeda(lucroPorArrobaProduzida(relatorio))}`);
    linhas.push(`Arroba de equilíbrio: ${moeda(precoArrobaEquilibrio(relatorio))}`);
    linhas.push(`Resultado só da engorda: ${moeda(margemDaEngorda(relatorio))}`);
    if (!(lote.precoCompra > 0)) {
      linhas.push("Sem preço de compra informado: o resultado conta apenas a dieta.");
    }
    linhas.push(
      "Não entram sanidade, transporte, pastagem, mão de obra nem impostos.",
    );
    linhas.push("");
  }

  linhas.push("PERÍODOS");
  for (const periodo of relatorio.periodos) {
    linhas.push(
      `${tituloPeriodo(periodo)}: ${formatarData(periodo.dataInicio)} a ` +
        `${formatarData(periodo.dataFim)} (${numero(periodo.dias, 0)} dias)`,
    );
    linhas.push(
      `  Peso ${numero(periodo.pesoInicial, 0)} a ${numero(periodo.pesoFinal, 0)} kg`,
    );
    linhas.push(
      `  Exigência diária por animal: MS ${formatarKg(periodo.exigencia.consumoMateriaSeca)}, ` +
        `PB ${gramas(periodo.exigencia.proteinaBrutaGramas)}, ` +
        `NDT ${formatarKg(periodo.exigencia.ndtKg)}`,
    );
    linhas.push(
      `  Dieta: ${percentual(periodo.exigencia.proteinaBrutaPercentualDieta)} PB e ` +
        `${percentual(periodo.exigencia.ndtPercentualDieta)} NDT na matéria seca`,
    );
    for (const consumo of periodo.consumos) {
      let texto = `  - ${consumo.insumo.nome}: ${numero(consumo.kgMateriaNatural, 0)} kg`;
      const conversao = consumoConversao(consumo);
      if (conversao) texto += ` = ${descricaoConversao(conversao)}`;
      linhas.push(texto);
    }
    const custo = custoPeriodo(periodo);
    if (custo > 0) linhas.push(`  Custo do período: ${moeda(custo)}`);
  }

  if (relatorio.alertas.length > 0) {
    linhas.push("");
    linhas.push("OBSERVAÇÕES");
    linhas.push(...relatorio.alertas.map((a) => `- ${a}`));
  }

  linhas.push("");
  linhas.push(
    "Gerado pelo NovilhaNutri. Cálculos de referência; ajuste conforme o desempenho " +
      "observado e a orientação do responsável técnico.",
  );
  return linhas.join("\n");
}
